crates/turborepo-scm/src/package_deps.rs

```diff
@@ -207,6 +242,45 @@ impl Git {
         hash_objects(&self.root, &full_pkg_path, to_hash, &mut hashes)?;
         Ok(hashes)
     }
+
+    #[tracing::instrument(skip(self, turbo_root, inputs))]
+    fn get_package_file_hashes_from_inputs_and_index<S: AsRef<str>>(
+        &self,
+        turbo_root: &AbsoluteSystemPath,
+        package_path: &AnchoredSystemPath,
+        inputs: &[S],
+        include_configs: bool,
+    ) -> Result<GitHashes, Error> {
+        // collect the default files and the inputs
+        let default_file_hashes =
+            self.get_package_file_hashes_from_index(turbo_root, package_path)?;
+
+        // we need to get hashes for excludes separately so we can remove them from the
+        // defaults later on
+        let mut includes = Vec::new();
+        let mut excludes = Vec::new();
+        for input in inputs {
+            let input_str = input.as_ref();
+            if let Some(exclude) = input_str.strip_prefix('!') {
+                excludes.push(exclude);
+            } else {
+                includes.push(input_str);
+            }
+        }
+        let manual_includes_hashes =
+            self.get_package_file_hashes_from_inputs(turbo_root, package_path, &includes, true)?;
+        let manual_excludes_hashes =
+            self.get_package_file_hashes_from_inputs(turbo_root, package_path, &excludes, false)?;
+
+        // merge the two includes
+        let mut hashes = default_file_hashes;
+        hashes.extend(manual_includes_hashes);
+
+        // remove the excludes
+        hashes.retain(|key, _| !manual_excludes_hashes.contains_key(key));
```

Comment:
Future work: I think there's some low hanging perf work to do in future PRs:
 - we can avoid calculating file hashes for all of the paths in `manual_excludes_hashes`
 - can also defer calculating all file hashes until we have the final list of what files we care about